search-app/src/cmr/search/services/query_service.clj

```diff
@@ -301,6 +304,102 @@
      (ph/provider-holdings->string
        (:result-format params) provider-holdings {:echo-compatible? (= "true" echo-compatible)})]))
 
+(defn- get-collections-with-deleted-revisions
+  "Executes elasticsearch searches to find collections that have deleted revisions.
+   Returns a list of concept-ids of the collections."
+  [context params]
+  ;;1/ Find all collection revisions that are deleted satisfy the original search params
+  (let [query (make-concepts-query context :collection params)
+        condition (gc/and-conds
+                   [(:condition query)
+                    (qm/boolean-condition :deleted true)])
+        results (common-idx/execute-query context
+                                          (-> query
+                                              (assoc :all-revisions? true)
+                                              (assoc :condition condition)
+                                              (assoc :page-size :unlimited)))
+        coll-concept-ids (map #(first (:concept-id (:fields %)))
+                              (get-in results [:hits :hits]))]
+    (distinct coll-concept-ids)))
+
+(defn- get-visible-collections
+  "Returns the concept ids of collections that are visible from the given collection concept ids"
+  [context coll-concept-ids]
+  (when (seq coll-concept-ids)
+    (let [condition (qm/string-conditions :concept-id coll-concept-ids true)
+          query (qm/query {:concept-type :collection
+                           :condition condition
+                           :page-size :unlimited
+                           :result-format :query-specified
+                           :result-fields []})
+          results (qe/execute-query context query)]
+      (map :concept-id (:items results)))))
+
+(defn- get-highest-visible-revisions
+  "Returns the query and the highest visible collection revisions search result of the given
+   collection concept ids in the given result format."
+  [context coll-concept-ids result-format]
+  (when (seq coll-concept-ids)
+    ;; find all collection revisions, then filter out the highest revisions
+    ;; and replace the hits and items in results with those of the highest revisions.
+    (let [condition (gc/and-conds
+                     [(qm/string-conditions :concept-id coll-concept-ids true)
+                      (qm/boolean-condition :deleted false)])
+          query (qm/query {:concept-type :collection
+                           :condition condition
+                           :all-revisions? true
+                           :page-size :unlimited
+                           :result-format result-format})
+          results (qe/execute-query context query)
+          highest-coll-revisions (u/map-values
+                                  #(apply max (map :revision-id %))
+                                  (group-by :concept-id (:items results)))
+          highest-revisions (filter
+                             #((set highest-coll-revisions)
```

Comment:
Probably easier to read if changing to (fn[] since it covers multiple lines. It took me a few reads to really get what this code was doing using set as a function to look to see if the given concept revision existed in the set.